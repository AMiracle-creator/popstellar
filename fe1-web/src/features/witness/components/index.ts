export {
  default as WitnessNotification,
  WitnessNotificationTypeComponent,
} from './WitnessNotification';
