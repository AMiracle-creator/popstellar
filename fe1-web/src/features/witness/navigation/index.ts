export { default as WitnessNavigation } from './WitnessNavigation';
