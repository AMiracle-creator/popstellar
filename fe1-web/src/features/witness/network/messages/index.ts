export * from 'features/witness/network/messages/WitnessMessage';
