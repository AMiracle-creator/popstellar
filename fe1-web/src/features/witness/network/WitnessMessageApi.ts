import { KeyPairStore } from 'store';
import { Channel, Hash } from 'core/objects';
import { publish } from 'core/network/jsonrpc/JsonRpcApi';

import { WitnessMessage } from './messages';

/**
 * Contains all functions to send witness related messages.
 */

/** Send a server messages to acknowledge witnessing the messages messages (JS object) */
export function requestWitnessMessage(channel: Channel, messageId: Hash): Promise<void> {
  const message = new WitnessMessage({
    message_id: messageId,
    signature: KeyPairStore.getPrivateKey().sign(messageId),
  });

  return publish(channel, message);
}
