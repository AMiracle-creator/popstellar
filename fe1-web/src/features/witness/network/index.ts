import { ActionType, MessageRegistry, ObjectType } from 'core/network/jsonrpc/messages';
import { handleWitnessMessage } from './WitnessHandler';
import { WitnessMessage } from './messages';

/**
 * Configures the network callbacks in a MessageRegistry.
 *
 * @param registry - The MessageRegistry where we want to add the mappings
 */
export function configureNetwork(registry: MessageRegistry) {
  registry.add(
    ObjectType.MESSAGE,
    ActionType.WITNESS,
    handleWitnessMessage,
    WitnessMessage.fromJson,
  );
}
