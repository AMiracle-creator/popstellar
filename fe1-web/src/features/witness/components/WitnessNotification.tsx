import PropTypes from 'prop-types';
import React, { FunctionComponent, useEffect, useMemo } from 'react';
import { Text, View } from 'react-native';
import { Button } from 'react-native-elements';
import { useSelector } from 'react-redux';

import { makeMessageSelector } from 'core/network/ingestion';
import { dispatch } from 'core/redux';

import { WitnessHooks } from '../hooks';
import { WitnessFeature } from '../interface';
import { requestWitnessMessage } from '../network/WitnessMessageApi';
import { removeMessageToWitness } from '../reducer';

const WitnessNotification = ({ notification, navigateToNotificationScreen }: IPropTypes) => {
  const messageSelector = useMemo(
    () => makeMessageSelector(notification.messageId),
    [notification.messageId],
  );
  const message = useSelector(messageSelector);

  const discardNotifications = WitnessHooks.useDiscardNotifications();
  const markNotificationAsRead = WitnessHooks.useMarkNotificationAsRead();
  const isEnabled = WitnessHooks.useIsEnabled();
  const laoId = WitnessHooks.useCurrentLaoId();

  const onWitness = () => {
    if (message) {
      dispatch(
        discardNotifications({ laoId: laoId.valueOf(), notificationIds: [notification.id] }),
      );
      if (isEnabled) {
        requestWitnessMessage(message.channel, message.message_id);
      }
      dispatch(removeMessageToWitness(message.message_id.valueOf()));
      navigateToNotificationScreen();
    }
  };

  const onDecline = () => {
    if (message) {
      dispatch(markNotificationAsRead({ laoId: laoId.valueOf(), notificationId: notification.id }));
      navigateToNotificationScreen();
    }
  };

  // if the notification state somehow gets out of sync, remove the corresponding notification
  useEffect(() => {
    if (!message) {
      dispatch(
        discardNotifications({ laoId: laoId.valueOf(), notificationIds: [notification.id] }),
      );
      navigateToNotificationScreen();
    }
  }, [laoId, navigateToNotificationScreen, discardNotifications, notification.id, message]);

  return (
    <View>
      <Text>{JSON.stringify(notification)}</Text>
      <Text>{JSON.stringify(message)}</Text>
      <Button title="Witness Message" onPress={onWitness} />
      <Button title="Decline Message" onPress={onDecline} />
    </View>
  );
};

const propTypes = {
  notification: PropTypes.shape({
    // WitnessFeature.MessageToWitnessNotification
    id: PropTypes.number.isRequired,
    timestamp: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    messageId: PropTypes.string.isRequired,
  }).isRequired,
  navigateToNotificationScreen: PropTypes.func.isRequired,
};
WitnessNotification.propTypes = propTypes;

type IPropTypes = PropTypes.InferProps<typeof propTypes>;

export default WitnessNotification;

export const WitnessNotificationType = {
  /**
   * Checks whether the given notification is a witness notification
   * @param notification The notification whose type should be checked
   * @returns True if the notification is a witness notification, false otherwise
   */
  isOfType: (notification: WitnessFeature.Notification) =>
    'type' in notification &&
    notification.type === WitnessFeature.NotificationTypes.MESSAGE_TO_WITNESS,

  /**
   * Custom cleanup function that removes the message from the witness store
   */
  delete: ((notification: WitnessFeature.MessageToWitnessNotification) => {
    dispatch(removeMessageToWitness(notification.messageId.valueOf()));
  }) as (notification: WitnessFeature.Notification) => void,

  /**
   * The component to render the witness notification
   */
  Component: WitnessNotification as unknown as FunctionComponent<{
    notification: WitnessFeature.Notification;
    navigateToNotificationScreen: () => void;
  }>,
};
