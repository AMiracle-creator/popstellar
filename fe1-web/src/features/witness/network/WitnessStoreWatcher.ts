import { Store } from 'redux';

import { KeyPairStore } from 'core/keypair';
import { getMessagesState } from 'core/network/ingestion';
import { ExtendedMessage } from 'core/network/ingestion/ExtendedMessage';
import { dispatch, getStore } from 'core/redux';

import {
  MESSAGE_TO_WITNESS_NOTIFICATION_TYPE,
  WitnessConfiguration,
  WitnessFeature,
} from '../interface';
import { addMessageToWitness, getMessageToWitness } from '../reducer';
import { WitnessingType, getWitnessRegistryEntry } from './messages/WitnessRegistry';
import { requestWitnessMessage } from './WitnessMessageApi';

/**
 * Is executed after a message has been successfully handled.
 * It handles the passive witnessing for messages and prepares
 * the application store for the act of manually witnessing
 * other messages
 */
export const afterMessageProcessingHandler =
  (
    enabled: WitnessConfiguration['enabled'],
    addNotification: WitnessConfiguration['addNotification'],
    getCurrentLao: WitnessConfiguration['getCurrentLao'],
    /* isLaoWitness: WitnessConfiguration['isLaoWitness'] */
  ) =>
  (msg: ExtendedMessage) => {
    // check if this message has already been signed by us
    const publicKey = KeyPairStore.getPublicKey();
    const signedByUs = msg.witness_signatures.find((sig) =>
      sig.signature.verify(publicKey, msg.message_id),
    );

    if (signedByUs) {
      // if it was, return. we do not have to sign it twice.
      return;
    }

    const storedMessage = getMessageToWitness(msg.message_id.valueOf(), getStore().getState());
    if (storedMessage) {
      // this message is already stored in the witness reducer
      // and hence does not have to be stored a second time
      return;
    }

    const entry = getWitnessRegistryEntry(msg.messageData);
    const lao = getCurrentLao();

    if (entry) {
      // we have a wintessing entry for this message type

      switch (entry.type) {
        case WitnessingType.PASSIVE:
          if (!enabled) {
            return;
          }

          requestWitnessMessage(msg.channel, msg.message_id);
          break;

        case WitnessingType.ACTIVE:
          // only send witness messages if we are a witness
          /* if (!isLaoWitness()) {
           break;
         } */

          dispatch(addMessageToWitness(new ExtendedMessage(msg).toState()));
          dispatch(
            addNotification({
              laoId: lao.id.valueOf(),
              title: `Witnessing required: ${msg.messageData.object}#${msg.messageData.action}`,
              timestamp: Timestamp.EpochNow().valueOf(),
              type: MESSAGE_TO_WITNESS_NOTIFICATION_TYPE,
              messageId: msg.message_id.valueOf(),
            } as WitnessFeature.MessageToWitnessNotification),
          );
          break;

        case WitnessingType.NO_WITNESSING:
        default:
          break;
      }
    }
  };

/** Creates a function that reacts to newly processed messages by calling
 * 'afterMessageProcessingHandler'
 * @remarks
 * Follows the implementation of makeMessageStoreWatcher but reacts to newly
 * processed messages instead of newly unprocessed ones Even it is tempting to
 * just check for changes in unprocessedIds, we should not rely on this function
 * being called after every dispatch(). Thus it is possible that a
 * message has skipped unprocessedIds and we have to check allIds as well
 * @param store The redux store
 * @returns The listener that can be passed to store.subscribe()
 */
export const makeWitnessStoreWatcher = (
  store: Store,
  afterProcessingHandler: (msg: ExtendedMessage) => void,
) => {
  let previousAllIds: string[] = [];
  let previousUnprocessedIds: string[] = [];

  let currentAllIds: string[] = [];
  let currentUnprocessedIds: string[] = [];
  return () => {
    const state = store.getState();
    const msgState = getMessagesState(state);
    const allIds = msgState?.allIds || [];
    previousAllIds = currentAllIds;
    currentAllIds = allIds;

    const unprocessedIds = msgState?.unprocessedIds || [];
    previousUnprocessedIds = currentUnprocessedIds;
    currentUnprocessedIds = unprocessedIds;

    if (
      previousAllIds.length === currentAllIds.length &&
      previousAllIds.every((value, index) => value === currentAllIds[index]) &&
      previousUnprocessedIds.length === currentUnprocessedIds.length &&
      previousUnprocessedIds.every((value, index) => value === currentUnprocessedIds[index])
    ) {
      // no change detected
      return;
    }

    // get all message ids that are part of currentAllIds
    for (const msgId of currentAllIds) {
      // and that are currently not part of unprocessedIds
      if (currentUnprocessedIds.includes(msgId)) {
        return;
      }
      // and that either have not been part of previousAllIds OR
      // have been part of previousUnprocessedIds
      if (previousAllIds.includes(msgId) && !previousUnprocessedIds.includes(msgId)) {
        return;
      }

      // i.e. all messages that have been processed
      // since the last call of this function

      afterProcessingHandler(ExtendedMessage.fromState(msgState.byId[msgId]));
    }
  };
};
