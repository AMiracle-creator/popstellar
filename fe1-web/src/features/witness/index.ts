import { WitnessNotificationTypeComponent } from './components';
import { WitnessConfiguration, WitnessInterface, WITNESS_FEATURE_IDENTIFIER } from './interface';
import { configureNetwork } from './network';
import { witnessReducer } from './reducer';

/**
 * Configures the witness feature
 *
 * @param configuration - The witness configuration object
 * @returns The interface the witness feature exposes
 */
export function configure(configuration: WitnessConfiguration): WitnessInterface {
  configureNetwork(configuration);

  return {
    identifier: WITNESS_FEATURE_IDENTIFIER,

    notificationTypeComponents: [WitnessNotificationTypeComponent],

    context: {
      useCurrentLao: configuration.useCurrentLao,
      addNotification: configuration.addNotification,
      discardNotification: configuration.discardNotification,
      markNotificationAsRead: configuration.markNotificationAsRead,
    },
    reducers: {
      ...witnessReducer,
    },
  };
}
