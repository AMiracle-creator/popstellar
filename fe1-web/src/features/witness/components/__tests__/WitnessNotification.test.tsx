import { configureStore } from '@reduxjs/toolkit';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { combineReducers } from 'redux';

import {
  configureTestFeatures,
  mockAddress,
  mockChannel,
  mockKeyPair,
  mockLao,
  mockLaoId,
} from '__tests__/utils';
import FeatureContext from 'core/contexts/FeatureContext';
import { addMessages, messageReducer } from 'core/network/ingestion';
import { ExtendedMessage } from 'core/network/ingestion/ExtendedMessage';
import { ActionType, MessageData, ObjectType } from 'core/network/jsonrpc/messages';
import { Hash, Timestamp } from 'core/objects';
import {
  addNotification,
  discardNotifications,
  markNotificationAsRead,
  notificationReducer,
} from 'features/notification/reducer';
import {
  WITNESS_FEATURE_IDENTIFIER,
  WitnessFeature,
  WitnessReactContext,
} from 'features/witness/interface';
import { MessageToWitnessNotification } from 'features/witness/objects/MessageToWitnessNotification';
import { addMessageToWitness, witnessReducer } from 'features/witness/reducer';

import WitnessNotification from '../WitnessNotification';

const mockMessageId = new Hash('some message id');

// otherwise .fromData won't work
configureTestFeatures();

const timestamp = new Timestamp(1607277600);

// set up mock store
const mockStore = configureStore({
  reducer: combineReducers({
    ...notificationReducer,
    ...witnessReducer,
    ...messageReducer,
  }),
});
const mockNotification = new MessageToWitnessNotification({
  id: 0,
  laoId: mockLaoId,
  title: 'a notification',
  timestamp: new Timestamp(0),
  type: WitnessFeature.NotificationTypes.MESSAGE_TO_WITNESS,
  hasBeenRead: false,
  messageId: mockMessageId,
});

const msg = ExtendedMessage.fromMessage(
  ExtendedMessage.fromData(
    {
      object: ObjectType.ROLL_CALL,
      action: ActionType.CREATE,
      name: 'rollcall1',
      description: 'a description',
      id: 'uO7c5qv5zenCd99Q8gLgBg0amZpjUOrmSkN7wAuZ-KM=',
      creation: timestamp,
      location: 'BC410',
      proposed_end: 1718888400,
      proposed_start: 1718884800,
    } as MessageData,
    mockKeyPair,
    mockChannel,
  ),
  mockAddress,
  mockChannel,
);

mockStore.dispatch(addMessages(msg.toState()));
mockStore.dispatch(addMessageToWitness(msg.message_id));
mockStore.dispatch(addNotification(mockNotification.toState()));

const contextValue = {
  [WITNESS_FEATURE_IDENTIFIER]: {
    enabled: true,
    useCurrentLaoId: () => mockLaoId,
    useConnectedToLao: () => true,
    addNotification: (notification) => mockStore.dispatch(addNotification(notification)),
    discardNotifications: (args) => mockStore.dispatch(discardNotifications(args)),
    markNotificationAsRead: (args) => mockStore.dispatch(markNotificationAsRead(args)),
    useCurrentLao: () => mockLao,
  } as WitnessReactContext,
};

describe('WitnessNotification', () => {
  it('renders correctly', () => {
    const component = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <WitnessNotification
            notification={mockNotification}
            navigateToNotificationScreen={() => {}}
          />
        </FeatureContext.Provider>
      </Provider>,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
});

describe('WitnessNotification actions', () => {
  let dispatchSpy: jest.SpyInstance;

  beforeEach(() => {
    // Spy on the store's dispatch function
    dispatchSpy = jest.spyOn(mockStore, 'dispatch');
  });

  afterEach(() => {
    // Restore the original function after each test
    dispatchSpy.mockRestore();
  });

  it('renders correctly for witnessing', () => {
    const { getByTestId } = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <WitnessNotification
            notification={mockNotification}
            navigateToNotificationScreen={() => {}}
          />
        </FeatureContext.Provider>
      </Provider>,
    );

    fireEvent.press(getByTestId('on-witness'));
    expect(dispatchSpy).toHaveBeenCalledWith(
      discardNotifications({ laoId: mockLaoId, notificationIds: [mockNotification.id] }),
    );
  });
  it('renders correctly for declining', () => {
    const { getByTestId } = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <WitnessNotification
            notification={mockNotification}
            navigateToNotificationScreen={() => {}}
          />
        </FeatureContext.Provider>
      </Provider>,
    );

    fireEvent.press(getByTestId('on-decline'));
    expect(dispatchSpy).toHaveBeenCalledWith(
      discardNotifications({ laoId: mockLaoId, notificationIds: [mockNotification.id] }),
    );
  });
});
