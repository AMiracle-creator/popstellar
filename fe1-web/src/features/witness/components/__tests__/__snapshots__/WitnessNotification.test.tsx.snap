// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WitnessNotification renders correctly 1`] = `
<View>
  <Text>
    {"id":0,"laoId":"-M259c5JlakLVlXBRa893pcE-4s7ssO8Jv5gXOvtreE=","hasBeenRead":false,"timestamp":0,"title":"a notification","type":"message-to-witness","messageId":"some message id"}
  </Text>
  <Text />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#3742fa",
            "borderColor": "#3742fa",
            "borderRadius": 8,
            "borderWidth": 1,
            "padding": 8,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
              "fontSize": 20,
              "lineHeight": 26,
              "textAlign": "left",
            },
            {
              "textAlign": "center",
            },
            {
              "color": "#fff",
            },
          ]
        }
      >
        Witness Message
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#3742fa",
            "borderColor": "#3742fa",
            "borderRadius": 8,
            "borderWidth": 1,
            "padding": 8,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
              "fontSize": 20,
              "lineHeight": 26,
              "textAlign": "left",
            },
            {
              "textAlign": "center",
            },
            {
              "color": "#fff",
            },
          ]
        }
      >
        Decline Message
      </Text>
    </View>
  </View>
</View>
`;
